linker/shared/src/main/scala/org/scalajs/linker/analyzer/Analyzer.scala

```diff
@@ -375,112 +375,152 @@ private class AnalyzerRun(config: CommonPhaseConfig, initial: Boolean,
       infos
     }
 
-    private def lookupClassForLinking(className: ClassName,
-        origin: LoadingClass): Future[LoadingResult] = {
-      ensureLoading(className, syntheticKind = None) match {
-        case loading: LoadingClass => loading.requestLink(origin)
-        case info: ClassInfo       => Future.successful(info)
-      }
-    }
-
     private def ensureLoading(className: ClassName,
         syntheticKind: Option[SyntheticClassKind]): ClassLoadingState = {
-      var loading: LoadingClass = null
+      var loading: LoadingInfos = null
       val state = _classInfos.getOrElseUpdate(className, {
-        loading = new LoadingClass(className)
+        loading = new LoadingInfos
         loading
       })
 
       if (state eq loading) {
         // We just added `loading`, actually load.
-        val result: Future[LoadingResult] = syntheticKind match {
+        syntheticKind match {
           case None =>
             val maybeInfo = infoLoader.loadInfo(className)
             val info = maybeInfo.getOrElse {
               Future.successful(createMissingClassInfo(className))
             }
-            info.flatMap { data =>
+            info.foreach { data =>
               doLoad(data, loading, syntheticKind, nonExistent = maybeInfo.isEmpty)
             }
 
           case Some(SyntheticClassKind.Lambda(descriptor)) =>
             val data = LambdaSynthesizer.makeClassInfo(descriptor, className)
             doLoad(data, loading, syntheticKind, nonExistent = false)
         }
-
-        loading.completeWith(result)
       }
 
       state
     }
 
-    private def doLoad(data: Infos.ClassInfo, origin: LoadingClass,
+    private def doLoad(data: Infos.ClassInfo, loadingInfos: LoadingInfos,
         syntheticKind: Option[SyntheticClassKind],
-        nonExistent: Boolean): Future[LoadingResult] = {
+        nonExistent: Boolean): Unit = {
       val className = data.className
 
-      for {
-        maybeAncestors <- Future.traverse(data.superClass.toList ++ data.interfaces)(
-            lookupClassForLinking(_, origin))
-      } yield {
-        val maybeCycle = maybeAncestors.collectFirst {
-          case cycle @ CycleInfo(_, null) => cycle
-
-          case CycleInfo(c, root) if root == className =>
-            CycleInfo(className :: c, null)
-
-          case CycleInfo(c, root) =>
-            CycleInfo(className :: c, root)
+      // Start loading the parents.
+      val parentNames = data.superClass.toList ++ data.interfaces
+      val parentsFuture = Future.traverse(parentNames) {
+        ensureLoading(_, syntheticKind = None) match {
```

Comment:
Consider naming the lambda argument here; it's a bit non-obvious:
```suggestion
      val parentsFuture = Future.traverse(parentNames) { parentName =>
        ensureLoading(parentName, syntheticKind = None) match {
```